File: golang/longest_word_in_dictionary_through_deleting.go
package golang

// Space	: O(1)
// Time		: O(n^2)

func findLongestWord(s string, d []string) string {
    ans := ""
    
    for _, word := range d {
        a, b := len(word), len(ans)
        if a < b || (a == b && word > ans) { 
            continue
        }
        
        i := 0
        for _, c := range s {
            if byte(c) == word[i] {
                i++
            }
            
            if i == a {
                ans = word
                break
            }   
        }
    }
    
    
    return ans
}

File: python3/longest_word_in_dictionary_through_deleting.py
"""
Space   : O(n)
Time    : O(n**2)
"""


class Solution:
    def findLongestWord(self, s: str, d: List[str]) -> str:
        ans = ""

        for word in d:
            a, b = len(word), len(ans)
            if a < b or (a == b and word > ans):
                continue

            i = 0
            for x in s:
                if x == word[i]:
                    i += 1
                if i == a:
                    ans = word
                    break

        return ans
